src/test/ui/consts/const-eval/write-to-uninhabited-enum-variant.rs

```diff
@@ -0,0 +1,30 @@
+// run-pass
+
+enum Empty { }
+enum Test1 {
+    A(u8),
+    B(Empty),
+}
+enum Test2 {
+    A(u8),
+    B(Empty),
```

Comment:
@RalfJung Kind of unrelated to this PR but I noticed when writing this test case, that the `Test2` enum doesn't trigger an ICE like `Test1` does. I believe this is because `Test` falls into this case here

https://github.com/rust-lang/rust/blob/8c7b921feb5b100b1825ad54d46a8cca575b35b3/src/librustc_mir/interpret/place.rs#L1038-L1040

but `Test2` falls into this case here

https://github.com/rust-lang/rust/blob/8c7b921feb5b100b1825ad54d46a8cca575b35b3/src/librustc_mir/interpret/place.rs#L1041-L1059

which only validates that the discriminant is in the valid range. I think it's missing a check similar to the one I added here 

https://github.com/rust-lang/rust/blob/f7cefd7a45980ad13e6fd4cf15ffdfa0bcfe9097/src/librustc_mir/transform/const_prop.rs#L564-L568

which verifies the variant isn't `Abi::Uninhabited`. It seemed weird to me that the behavior wasn't consistent.

Reply:
I'm not suggesting we special case anything; rather the opposite. Perhaps my initial assumption was wrong so let me back up a bit.

While tracking down the CI failure, I came up with this test case which causes an ICE:

```rust
enum Empty { }
enum Test {
    A(u8),
    B(Empty),
}

fn bar() -> Option<Empty> {
    std::process::exit(0)
}

fn main() {
    if let Some(x) = bar() {
        Test::B(x);
    }
}
```

based on our conversation yesterday, I think we decided it would be best to leave the `assert_eq!()` in. 

Last night, while working on a fix that wouldn't change miri, I discovered that if I simply added another case to `Test` and changed nothing else, the test would no longer ICE. 

So, to me, `Test::B(x)` causing an ICE when `Test` has two variants but not when it has three is weird. Do you agree with that?

Reply:
Did you add another uninhabited variant, or an inhabited variant.

Reply:
Inhabited (I think), for example:

```rust
enum Test {
    A(u8),
    B(Empty),
    C,
}
```

causes the ICE to disappear.

Reply:
In that case `Variants::Multiple` gets used instead of `Variants::Single`. This means that the assertion is never reached, as it is only used for `Variants::Single`.

Reply:
Yeah, that's what I was alluding to in https://github.com/rust-lang/rust/pull/64890#discussion_r333540295. That is what seems inconsistent to me. My (perhaps naive) opinion is `Test::B(x)` should either always assert or never assert. It shouldn't matter how many variants `Test` has, the behavior of the evaluator on `Test::B(x)` should be consistent.

Reply:
> So, to me, Test::B(x) causing an ICE when Test has two variants but not when it has three is weird. Do you agree with that?

Layout code is subtle, and probably these two types "feel" very different to the evaluator. So I am not entirely surprised.

Single-variant enums (including enums where all but one variant are uninhabited) are basically structs. They don't even have a discriminant at run-time! So I would not necessarily expect to see all the same behavior.

> I'm not suggesting we special case anything

Well I am seeing a `if variant_layout.abi == Abi::Uninhabited {` in the code you linked. That is the very definition of adding a special case, and I think it is wrong. If you ever need a conditional like that, something somewhere else is not handling enums properly.

Reply:
> Last night, while working on a fix that wouldn't change miri, I discovered that if I simply added another case to Test and changed nothing else, the test would no longer ICE.

> So, to me, Test::B(x) causing an ICE when Test has two variants but not when it has three is weird. Do you agree with that?

The question is, what's the MIR? If in one case the MIR sets an invalid discriminant and in the other case it doesn't, I'd say the interpreter is perfectly consistent. MIR building is not, but we should not paper over that on later phases.

Reply:
If it's ok with you @RalfJung, why don't we sync up on Zulip 

https://rust-lang.zulipchat.com/#narrow/stream/189540-t-compiler.2Fwg-mir-opt/topic/.5Bconst-prop.5D.20Handle.20remaining.20MIR.20Rvalue.20cases.20.2364890

Reply:
Maybe tomorrow; today I'm afraid I don't have time for a synchronous conversation.